File: atcoder/abc/074/D.py
def main():
    """
    ワーシャルフロイドで経由したときに変更がないことを確認する
    変更がある場合、
        最短経路ではないため条件を満たさない
    変更がない場合、
        経由の経路でも最短距離になるケースを除いて距離を計算する
    """
    N = int(input())
    a = [None] * N
    total = 0
    for i in range(N):
        a[i] = list(map(int, input().split()))
        total += sum(a[i])

    for k in range(N):
        for i in range(N):
            for j in range(N):
                # compare from-to / from-経由地点-to
                d = a[i][k] + a[k][j]
                if a[i][j] > d:
                    return -1

    # 不要判定
    v = [[False] * N for _ in range(N)]
    no_need = 0
    for k in range(N):
        for i in range(N):
            for j in range(N):
                if i == k or j == k:
                    continue
                if not v[i][j] and (a[i][j] == a[i][k] + a[k][j]):
                    no_need += a[i][j]
                    v[i][j] = True

    # 対称なので半分にする
    return (total - no_need) // 2

if __name__ == '__main__':
    answer = main()
    print(answer)

File: atcoder/abc/074/D.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	// http://qiita.com/tnoda_/items/b503a72eac82862d30c6
	var N int
	fmt.Scan(&N)

	// make 2 dimension slice
	a := make([][]int, N)
	for i := 0; i < N; i++ {
		a[i] = make([]int, N)
	}

	// 行取得からの分割/Int
	var sc = bufio.NewScanner(os.Stdin)
	total := 0
	for i := 0; i < N; i++ {
		sc.Scan()
		line := sc.Text()
		values := strings.Split(line, " ")
		for j, v := range values {
			vInt, _ := strconv.Atoi(v)
			a[i][j] = vInt
			total += vInt
		}
	}

	no_need := 0
	for k := 0; k < N; k++ {
		// from
		for i := 0; i < N; i++ {
			// to
			for j := 0; j < N; j++ {
				// compare from-to / from-経由地点-to
				if i == k || j == k {
					continue
				}
				d := a[i][k] + a[k][j]
				if a[i][j] > d {
					fmt.Println(-1)
					return
				} else if a[i][j] == d {
					// 経由して同じであれば必要ない
					no_need += d
				}
			}
		}
	}
	//  対称なので半分にする
	fmt.Println((total - no_need) / 2)
}
